Feature get all revision

AleXKarasev opened the pull request:
Сделал 3 комита:
1. (0982b90) добавил обработку GET api/counter (admin, editor, anonymous) - возвращающий все каунтеры
2. (092a1c7) поправил багу #5 (новый каунтер возвращает 0). Поправил тест и добавил новый
3. (14f8530) обновил пакеты Nuget до последней версии, убрал binginRedirect в конфиге

AleXKarasev committed:
Добавил дефолтный метод контроллера, возвращающий все каунтеры

AleXKarasev committed:
При создании нового каунтера возвращается 0

AleXKarasev committed:
Обновил референсы до последней версии, убрал binginRedirect

ExM commented on BuildRevisionCounter.Tests/Controllers/CounterControllerTest.cs:
```diff
@@ -43,12 +44,20 @@ public void SetUp()
+            const string revName = "BumpingNewRevisionMayBeInvokedMultipleTimes";
+            await _controller.Bumping(revName);
+            await _controller.Bumping(revName);
+        }
```
Условия прохождения теста слишком широкие, т.к. в нем нет ни одного утверждения.

ExM commented on BuildRevisionCounter.Tests/Controllers/CounterControllerTest.cs:
```diff
@@ -43,12 +44,20 @@ public void SetUp()
+			Assert.AreEqual(0, rev);
 		}
 
+        [Test]
```
Добавил требования по форматированию кода.
Мы используюем [RSDN](http://rsdn.ru/article/mag/200401/codestyle.XML)
Здесь - отступы следует делать через TAB

AleXKarasev replied:
Согласен, увидел что мое форматирование отличается уже в момент отправки пулреквеста, исправлю

ExM commented on BuildRevisionCounter/Controllers/CounterController.cs:
```diff
@@ -27,6 +28,21 @@ public CounterController(IMongoDBStorage mongoDbStorage)
+        [HttpGet]
+        [Route("")]
+        [Authorize(Roles = "admin, editor, anonymous")]
+        public async Task<IReadOnlyCollection<RevisionModel>> GetAllRevision()
```
Этот запрос с непредсказуемым временем выполнения.
Какие есть варианты ограничить время и(или) память при выполнении этого запроса?

AleXKarasev replied:
Согласен, упустил что каунтеров может быть большое количество. Возможные решения проблемы: отдавать каунтеры порциями, размер порции и ее номер принимать через параметры. По умолчанию (если параметры запроса не указаны) отдавать n первых

ExM commented on BuildRevisionCounter/Controllers/CounterController.cs:
```diff
@@ -47,21 +63,36 @@ public CounterController(IMongoDBStorage mongoDbStorage)
-					});
+            var existRevision = await _mongoDbStorage.Revisions
+                .CountAsync(r => r.Id == revisionName);
+		    if (existRevision == 0)
```
- На каждый клиентский запрос теперь требуется 2 обращения к БД. Можно ли ограничиться одним?
- После первого запроса состоянии коллекцйии может быть изменено. Запрос в ветке кода с InsertOneAsync может упасть с ошибкой.

AleXKarasev replied:
Да тоже хотел обойтись 1 запросом который сразу все будет делать, но не нашел решения как это сделать. Поэтому оставил вариант с двумя запросами. Возможно подойдет решение с использованием eval и передачей js скрипта в базу, но я достаточно мало знаком с монгой

ExM commented:
После обновления пакетов приложение не запускается с ошибкой:

```
Не удалось загрузить файл или сборку "System.Web.Http, Version=5.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" либо одну из их зависимостей. Найденное определение манифеста сборки не соответствует ссылке на сборку. (Исключение из HRESULT: 0x80131040)
```

AleXKarasev commented:
Странно у меня после обновления пакетов ошибки не возникало, попробую переклонировать репозиторий

AleXKarasev committed:
Поправил замечания:

	1. Добавил пагинацию для метода возвращающего все каунтеры
	2. Добавил блокировку на создание нового каунтера. Уйти от 2 запросов к БД не получилось. Проблему многопоточного доступа решил в коде сервиса

AleXKarasev committed:
Поправил табуляцию у теста

AleXKarasev commented:
Поправил замечания. Проблема с компиляцией не смоделировалась. Переклонировал реп, сделал резолв зависисмостей, все собралось без ошибок.

ExM commented on BuildRevisionCounter.Tests/Controllers/CounterControllerTest.cs:
```diff
@@ -43,10 +45,25 @@ public void SetUp()
+			}
+			catch (Exception)
+			{
+				Assert.Fail("Ошибка при создании одного каунтера из разных потоков");
```
Тест выполняет два обновления счетчика. Один за другим. В одном или разных потоках.
Уборка кода в try/catch приводит к потере стека в сообщениях NUnit.

Здесь надо добавить проверку, что номер ревизии увеличился на 2, т.к. просто отсутствие исключения не говорит о том что код делает, то что должен.

Кстати, имя счетчика каждый раз можно генерить новое (GUID) и быть всегда уверенным, что это отсутствующий счетчик без зачистки БД.

ExM commented on BuildRevisionCounter/Controllers/CounterController.cs:
```diff
@@ -42,25 +63,48 @@ public CounterController(IMongoDBStorage mongoDbStorage)
+			// накладывааем блокировку на проверку каунтера 
+			// и вставку начального значения если каунтера еще нет
+			lock (_sync)
+			{
```
В общем случае - плохое решение.
Данный сервис, скорее всего будет запущен в одном экземпляре.
Но, если он работает в условиях балансировки запросов, то эта блокировка останется на одном сервере.

ExM commented on BuildRevisionCounter/Controllers/CounterController.cs:
```diff
@@ -42,25 +63,48 @@ public CounterController(IMongoDBStorage mongoDbStorage)
 						.Set(r => r.Updated, DateTime.UtcNow),
 					new FindOneAndUpdateOptions<RevisionModel>
 					{
 						IsUpsert = true,
```
1. Если здесь выставить IsUpsert = false, то по результату запроса можно будет определить, что счетчик отсутствует в БД.
2. В случае отсутствия счетчика можно сделать вставку (с нулевым значением), если вставка не удалась (счетчик успели вставить раньше), то следует перейти к п.1

При обновлении счетчика будет всегда 1 запрос, при добавлении нового будет в основном 2 запроса.

ExM commented:
Проблем с компиляцией нет. Проблемы с запуском приложения. Чтобы все загрузилось требуется добавить еще 3 редиректа:
- System.Web.Http.Owin 5.2.3.0
- System.Web.Http 5.2.3.0
- Microsoft.Owin 3.0.1.0

ExM commented on BuildRevisionCounter/Controllers/CounterController.cs:
```diff
@@ -28,6 +29,26 @@ public CounterController(IMongoDBStorage mongoDbStorage)
+				throw new HttpResponseException(HttpStatusCode.NotFound);
+
+			return revisions;
+		}
```
А если... клиент укажет pageSize = 2147483647 ?

ExM commented:
Напоминаю - до понедельника ожидаю получить исправления.

AleXKarasev committed:
Переделал тест многопоточной работы. Добавил проверку конечного результата. Количество потоков можно изменять константой

AleXKarasev committed:
Переделал механизм обновления каунтера.

1. Сначала пробуем обновить с выключенным IsUpsert. Если приходит null значит записи с таким именем еще нет
2. Пытаемся вставить ее, если получилось возвращаем 0
3. Если при вставке произошла ошибка, то значит мы не успели вставить, запускаем опять обновление

AleXKarasev committed:
Удалил неиспользуемую переменную

AleXKarasev committed:
Редиректы в конфиге появляються из за Ninject (Ninject.Web.WebApi, Ninject.Web.WebApi.OwinHost, Ninject.Web.Common.OwinHost), он собран со старыми версиями пакетов System.Web.Http.Owin 5.0.0.0, System.Web.Http 5.0.0.0, Microsoft.Owin Microsoft.Owin 3.0.0.0.

Поэтому что бы не было биндинг редиректов эти пакеты не должны быть выше 5.0.0 версии. Так же после этого появилась проблема с 6 версий Newtonsoft.Json. System.Net.Http.Formatting и System.Web.Http собраны под 4,5 версию. В итоге Newtonsoft.Json должна быть не выше 4.5. Сборка, тесты и запуск проходят успешно

AleXKarasev committed:
Т.к. в монге нет таймаута на операцию, то сделал запуск параллельно работающих потоков, один из которых выполняет роль таймаута. Таймаут установил в 30 сек. Можно его вынести в параметры метода, но тогда можно будет опять сформировать запрос который приведёт к длительному выполнению метода. После окончания работы одного из потоков, посылается команда на отмену оставшегося.

AleXKarasev committed:
Нужно было добавить Microsoft.Owin.Host.SystemWeb, это реализация Owin для iis. И снова появился биндинг редирект из за старых сборок Ninject

AleXKarasev committed:
Добавил подписку на резолвы библиотек доменом. Что бы подписка вызывалась раньше всего, сделал ее в статическом конструкторе.

Таким образом удалось обновить все пакеты nuget до последней версии, кроме одного - System.Web.Http. При его резолве через код падает конструктор StandardKernel
В итоге все пакеты обновлены до последнего и остался один биндинг редирект

ExM commented on BuildRevisionCounter.Tests/Controllers/CounterControllerTest.cs:
```diff
@@ -43,10 +46,30 @@ public void SetUp()
+			List<Task> taskList = new List<Task>();
+			for (int i = 0; i < n; i++)
+			{
+				taskList.Add(Task.Run(async () => await _controller.Bumping(revName)));
```
Напрасно усложненный код. Должно быть так:

```
taskList.Add(_controller.Bumping(revName));
```

ExM commented on BuildRevisionCounter.Tests/Controllers/CounterControllerTest.cs:
```diff
@@ -64,5 +87,76 @@ public void SetUp()
+		{
+			// пересоздадим БД перед этим тестом,
+			// он должен проводиться на чистой БД
+			SetUpAsync().Wait();
```
Блокирующее ожидание в асинхронном коде.

ExM commented on BuildRevisionCounter/Controllers/CounterController.cs:
```diff
@@ -28,6 +30,39 @@ public CounterController(IMongoDBStorage mongoDbStorage)
+		public async Task<IReadOnlyCollection<RevisionModel>> GetAllRevision([FromUri] Int32 pageSize=20, [FromUri] Int32 pageNumber=1)
+		{
+			if (pageSize < 1 || pageNumber < 1)
+				throw new HttpResponseException(HttpStatusCode.BadRequest);
```
Надо просто граничить максимальный размер страницы до 1000 например.

ExM commented on BuildRevisionCounter/Controllers/CounterController.cs:
```diff
@@ -28,6 +30,39 @@ public CounterController(IMongoDBStorage mongoDbStorage)
+			if (pageSize < 1 || pageNumber < 1)
+				throw new HttpResponseException(HttpStatusCode.BadRequest);
+
+			CancellationTokenSource cts = new CancellationTokenSource();
```
CancellationToken можно разместить в аргументе метода.
А для ограничения времени работы метода реализовать фильтр.
Пример есть здесь: http://stackoverflow.com/questions/24403832/timeout-a-web-api-request

ExM commented on BuildRevisionCounter/Controllers/CounterController.cs:
```diff
@@ -47,6 +82,42 @@ public CounterController(IMongoDBStorage mongoDbStorage)
+				}
+				catch (MongoWriteException)
+				{
+					// если прои вставке произошла ошибка значит мы не успели и запись там уже есть
```
MongoWriteException может произойти по разным причинам. Пропущено определение нарушение unique индекса.

ExM commented on BuildRevisionCounter/Controllers/CounterController.cs:
```diff
@@ -47,6 +82,42 @@ public CounterController(IMongoDBStorage mongoDbStorage)
+				{
+					// если прои вставке произошла ошибка значит мы не успели и запись там уже есть
+					// и теперь попытка обновления должна пройти без ошибок
+					result = FindOneAndUpdateRevisionModelAsync(revisionName).Result;
```
Блокирующее ожидание в асинхронном методе.
В данном случае, если поток сюда доберется, то он тут зависнет.
Почему?

ExM commented on BuildRevisionCounter/Startup.cs:
```diff
@@ -1,24 +1,58 @@
+				return Assembly.Load(name);
+			}
+			return null;
+		}
```
Это даже хуже чем биндинг редиректы. Так хоть VS предупреждать могла о конфликтах. Теперь обновление пакетов до новой версии будет причудливо переплетено с правками в C# коде.

Следует отселить переезд на новые версии библиотек в отдельную ветку.

ExM commented on BuildRevisionCounter.Tests/Controllers/CounterControllerTest.cs:
```diff
@@ -43,10 +46,30 @@ public void SetUp()
+		[Test]
+		public async Task BumpingNewRevisionMayBeInvokedMultipleTimes()
+		{
+			string revName = "BumpingNewRevisionMayBeInvokedMultipleTimes" + Guid.NewGuid().ToString();
```
Почему это решение не распространено на остальные случаи?
